File: introductory_problems/bit_strings/BitStrings.java
// Working program using BufferedReader
import java.io.*;
import java.util.*;

public class BitStrings {
    private static long MOD = 1000000007;
    public static void main(String[] args) throws IOException {
        Scanner input = new Scanner(System.in);
        int n = Integer.valueOf(input.nextInt());
        input.close();
        Writer writer = new PrintWriter(System.out);
        long[] saved = new long[n + 1];
        saved[0] = 1;
        saved[1] = 2;
        long res = findNum(n, saved);
        writer.write(String.valueOf(res));
        writer.flush();
    }

    private static long findNum(int n, long[] saved){
        if (saved[n] != 0) return saved[n];
        else {
            if (n % 2 == 0) {
                long x = findNum(n / 2, saved);
                saved[n] = x * x % MOD;
                return saved[n];
            } else {
                long x = findNum(n / 2, saved);
                saved[n] = x * x * 2 % MOD;
                return saved[n]; 
            }
        }

    }


}

File: introductory_problems/bit_strings/BitStrings.py
def _get_val(k, saved_dict):
    if k in saved_dict:
        return saved_dict[k]
    else:
        half_val = _get_val(k // 2, saved_dict)
        if k % 2 == 0:
            saved_dict[k] = half_val * half_val % 1000000007
        else:
            saved_dict[k] = half_val * half_val * 2 % 1000000007
        return saved_dict[k]

n = int(input())

saved_dict = {}
saved_dict[0] = 1
saved_dict[1] = 2
res = _get_val(n, saved_dict)
print(res)